src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -77,110 +77,88 @@ public IEnumerable<AssemblyInfo> GetCachedAssemblyInfos(string fileName) => _nam
 
         public string GetCachedReferencePath(MetadataReferenceInfo referenceInfo)
         {
-            if (_mvidMap.TryGetValue(referenceInfo.Mvid, out var value))
+            if (_mvidMap.TryGetValue(referenceInfo.ModuleVersionId, out var value))
             {
                 return value.FilePath;
             }
 
             throw new Exception($"Could not find referenced assembly {referenceInfo}");
         }
 
-        public bool TryResolveReferences(ImmutableArray<MetadataReferenceInfo> references, out ImmutableArray<MetadataReference> results)
+        public bool TryResolveReferences(MetadataReferenceInfo metadataReferenceInfo, [NotNullWhen(true)] out MetadataReference? metadataReference)
         {
-            if (!CacheNames(references))
+            if (!TryGetAssemblyInfo(metadataReferenceInfo, out var assemblyInfo))
             {
-                results = default;
+                metadataReference = null;
                 return false;
             }
 
-            var builder = ImmutableArray.CreateBuilder<MetadataReference>(references.Length);
-            foreach (var reference in references)
-            {
-                var filePath = GetCachedReferencePath(reference);
-                using var fileStream = File.OpenRead(filePath);
+            using var fileStream = File.OpenRead(assemblyInfo.FilePath);
 
-                // This is deliberately using an ordinal comparison here. The name of the assembly is written out 
-                // into the PDB. Rebuild will only succeed if the provided reference has the same name with the
-                // same casing
-                if (Path.GetFileName(filePath) != reference.Name)
-                {
-                    filePath = Path.Combine(Path.GetDirectoryName(filePath)!, reference.Name);
-                }
-                builder.Add(MetadataReference.CreateFromStream(
-                    fileStream,
-                    filePath: filePath,
-                    properties: new MetadataReferenceProperties(
-                        kind: MetadataImageKind.Assembly,
-                        aliases: reference.ExternAlias is null ? ImmutableArray<string>.Empty : ImmutableArray.Create(reference.ExternAlias),
-                        embedInteropTypes: reference.EmbedInteropTypes)));
+            // This is deliberately using an ordinal comparison here. The name of the assembly is written out 
+            // into the PDB. Rebuild will only succeed if the provided reference has the same name with the
+            // same casing
+            var filePath = assemblyInfo.FilePath;
+            if (Path.GetFileName(filePath) != metadataReferenceInfo.FileName)
+            {
+                filePath = Path.Combine(Path.GetDirectoryName(filePath)!, metadataReferenceInfo.FileName);
             }
 
-            results = builder.MoveToImmutable();
+            metadataReference = MetadataReference.CreateFromStream(
```

Comment:
This method will read the whole file in and close the stream, so consider inlining the `fileStream` argument.<div><br></div><div class="">My radar goes off when we `using var` something and then directly or indirectly return that thing. We need to make sure that either we know the returned thing didn't capture the disposable resource, OR that it captured it and now owns the disposable resource, in which case we don't want to use `using` on the resource.</div><div><br></div>